### PR TITLE
Fix issue 18919 - __FILE__ and __LINE__ should work when used in defa…

### PR DESCRIPTION
**Pull request by tim-dlang**
…ult argument expressions

The parser now always creates AST nodes for default init expressions like `__FILE__`. They are replaced in resolveLoc. Variable inDefaultArg in Scope is used, so the nodes are not replaced too early.

### CODE REVIEW

File: changelog/dmd.default-init.dd [hunk 1]
```diff
@@ -0,0 +1,35 @@
+Keywords like `__FILE__` are evaluated at the call site more often
```
Comment (SixthDot) [inline changelog/dmd.default-init.dd:1]: more reliably

File: changelog/dmd.default-init.dd [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Dafault arguments for functions can contain the keywords `__FILE__`,
```
Comment (SixthDot) [inline changelog/dmd.default-init.dd:3]: Default

File: changelog/dmd.default-init.dd [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    printf("%s:%zd\n", file, line);
+}
+
+struct Loc
+{
+   string file;
+   size_t line;
+}
+
+void func2(Loc loc = Loc(__FILE__, __LINE__))
```
Comment (SixthDot) [inline changelog/dmd.default-init.dd:21]: you should mention that this case in particular did not work

File: changelog/dmd.default-init.dd [hunk 4]
```diff
@@ -0,0 +1,35 @@
+Keywords like `__FILE__` are evaluated at the call site more often
+
+Dafault arguments for functions can contain the keywords `__FILE__`,
+`__FILE_FULL_PATH__`, `__MODULE__`, `__LINE__`, `__FUNCTION__`
+and `__PRETTY_FUNCTION__`. They are now evaluated at the source location
+of the calling function in more complex expressions and not just when
```
Comment (SixthDot) [inline changelog/dmd.default-init.dd:6]: as long as used in an initializer, directly or not

**Review by ghost — CHANGES_REQUESTED** [review]
The changelog entry needs to be more accurate.

**Comment by tim-dlang** [comment]
The new test was not portable, because the output contained path separators and size_t was different. Now the output is more portable.

I also fixed another edge case: CatExp is lowered to `_d_arraycatnTX`, but `__FILE__` inside the lowering was not changed. Function resolveLoc now also runs for the lowering.

**Comment by tim-dlang** [comment]
Pipeline "Windows_VisualD_LDC x86-mscoff_MinGW" seems to not support %zd for printf, so the test now only uses %d and int for line numbers.

A memory allocation failure on "Windows_VisualD_LDC x64_Debug" seems to also randomly happen on other PRs. I have created an issue for that: https://issues.dlang.org/show_bug.cgi?id=24309

**Comment by schveiguy** [comment]
So cool, thanks!

**Comment by dkorpel** [comment]
What happens now if you do `f(int x = someTemplate!__LINE__)`?

File: compiler/src/dmd/expressionsem.d [hunk 5]
```diff
@@ -7475,7 +7475,11 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             printf("MixinExp::semantic('%s')\n", exp.toChars());
         }
 
+        const saveInDefaultArg = sc.inDefaultArg;
+        sc.inDefaultArg = false;
+
         auto e = compileIt(exp);
+        sc.inDefaultArg = saveInDefaultArg;
```
Comment (maxhaton) [inline compiler/src/dmd/expressionsem.d:7482]: It's not clear to me what's going on here

File: compiler/src/dmd/expressionsem.d [hunk 6]
```diff
@@ -15076,21 +15189,30 @@ Expression resolveLoc(Expression exp, const ref Loc loc, Scope* sc)
         return e.expressionSemantic(sc);
     }
 
-    Expression visitLineInit(LineInitExp _)
+    Expression visitLineInit(LineInitExp exp)
     {
+        if (sc.inDefaultArg)
+            return exp;
```
Comment (maxhaton) [inline compiler/src/dmd/expressionsem.d:15195]: This patterns is repeated quite a lot, can that be avoided?

File: compiler/test/runnable/issue18919.d [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+imports.issue18919b.func7: expr1=1078, file=issue18919.d func=issue18919.main, expr3=1
+imports.issue18919b.func8: expr1=[40, 1040, ], expr2=[issue18919.d: 40, ], expr3=constant2, expr4=issue18919
+imports.issue18919b.func9.fp: issue18919b.d:216 imports.issue18919b
+imports.issue18919b.func10: expr1=imports.issue18919b, expr2=imports.issue18919b
+---
+*/
+import imports.issue18919b;
+
+void main()
+{
```
Comment (maxhaton) [inline compiler/test/runnable/issue18919.d:26]: Could this be a `compilable`? (They're much faster).

**Comment by tim-dlang** [comment]
> What happens now if you do `f(int x = someTemplate!__LINE__)`?

Unfortunately that is now an error:
```
Error: cannot interpret `__LINE__` at compile time
```
I did not check this previously. It would be possible to change it, so it still compiles, but it would be the line at the template instantiation and not the call site.

File: compiler/src/dmd/expressionsem.d (hunk 5 above)
Comment (tim-dlang) [reply compiler/src/dmd/expressionsem.d:7482]: This is relevant for the following test:
```
void func(string expr = mixin("\"expr2=" ~ __MODULE__ ~ "\""))
{}
```
The mixin expression is inside a default argument, so `sc.inDefaultArg` is `true`. The code now sets `sc.inDefaultArg` temporarily to `false`, so the inside of the expression can be evaluated at compile time.
It's the same problem as for the example by @dkorpel with a template, but I only fixed it for a mixin expression and not templates.

File: compiler/src/dmd/expressionsem.d (hunk 6 above)
Comment (tim-dlang) [reply compiler/src/dmd/expressionsem.d:15195]: Good idea. This can be done once at the start of resolveLoc.

File: compiler/test/runnable/issue18919.d (hunk 7 above)
Comment (tim-dlang) [reply compiler/test/runnable/issue18919.d:26]: No, the test checks the output of the program at runtime und `RUN_OUTPUT`.

**Comment by schveiguy** [comment]
If you aren't going to "fix" the template thing, then the current behavior should be preserved. It should not become an error.

File: compiler/test/runnable/issue18919.d (hunk 7 above)
Comment (maxhaton) [reply compiler/test/runnable/issue18919.d:27]: A lot of the compilables use a static assert to check values.

**Comment by tim-dlang** [comment]
The example `f(int x = someTemplate!__LINE__)` now works again, but uses the value at the location of the template instantiation.

File: compiler/test/runnable/issue18919.d (hunk 7 above)
Comment (tim-dlang) [reply compiler/test/runnable/issue18919.d:27]: The values are not always available at compile time. Only `func5` and `func6` pass the values as template parameters, so they could be checked with static assert. Most of the test functions pass the values using runtime parameters, so a test at runtime is still necessary.

**Comment by pbackus** [comment]
The fact that we have a special case in the language for `__FILE__` and `__LINE__` as default arguments is not a good thing, and expanding this special case further is a move in the wrong direction. See the discussion in PR #14549 starting from [this comment][1] for a more detailed explanation.

I think a better way to support these use-cases would be to add a general-purpose mechanism for evaluating default arguments at the call site, which is not tied specifically to the presence of the `__FILE__` or `__LINE__` tokens.

[1]: https://github.com/dlang/dmd/pull/14549#issuecomment-1277444672

**Comment by schveiguy** [comment]
> I think a better way to support these use-cases would be to add a general-purpose mechanism for evaluating default arguments at the call site, which is not tied specifically to the presence of the `__FILE__` or `__LINE__` tokens.

I don't disagree, but a couple questions here:

1. Is this easy to do?
2. Is this easy to do later if we merge this now?

I don't want perfect to be the enemy of good here. We can make the code look nice later, right?

This is a real use case that can immediately be useful, since 99.999% of people create exceptions without specifying explicitly the line and file, they let the default arguments do their thing. We can easily deprecate the size_t and string parameters, and use some LOC struct to fix this, and then it secures infinitely more the intent of the user when building exceptions using explicit file/line parameters. I would hate to delay it because the implementation details are not perfect.

**Comment by pbackus** [comment]
> 1. Is this easy to do?

I don't think it's a huge challenge technically, but it probably requires a DIP

> 2. Is this possible to do later if we merge this now?

Yes, but then we're stuck with the special case.

That said, there's a case to be made that having `__FILE__` and `__LINE__` always behave like this is less of a special case then having them only behave that way at the top level, and I can't deny that it's useful. So maybe we should merge this now, and look into replacing it with a more general solution in a future language edition.

File: compiler/test/runnable/issue18919.d (hunk 7 above)
Comment (tim-dlang) [reply compiler/test/runnable/issue18919.d:27]: Maybe it would be possible to run everything at CTFE. Then it could be in `compilable`. But then it would not test that for example `__FILE__.ptr` is correctly handled by the backend.

**Comment by tim-dlang** [comment]
Currently, the semantic analysis for default arguments runs once at the declaration site. Later the AST for a default argument is copied to the call site and special tokens are replaced. This is not only done for the special tokens, but also for code like `f(Class obj = new Class)`, where a new object is created for every call, or `f(void* buffer = alloca(200))`, where memory is allocated on the stack, which has to be valid for the calling function. This even works with more complex expressions.

The compiler has another special case before this PR for function `void f(const(char)* file=__FILE__.ptr, const(char)* func=__FUNCTION__.ptr)
`. Here `__FUNCTION__` is evaluated at the call site, but `__FILE__` is evaluated at the declaration site. My PR implements all special tokens the same, so they are consistent.

A more general solution could be to run the semantic analysis for default arguments not just once, but for every call of a function. Symbol resolution should still be done in the context of the function declaration. Templates could then be instantiated with different values for special tokens as template arguments. This could also be more maintainable, because resolveLoc would not be needed anymore. Currently, resolveLoc would need to be updated for some changes to AST classes. On the other hand, running semantic analysis multiple times could be slower.

**Comment by PetarKirov** [comment]
> A more general solution could be to run the semantic analysis for default arguments not just once, but for every call of a function. [..]

This indeed sounds like a more general and principled approach to me.

> [..] On the other hand, running semantic analysis multiple times could be slower.

This worries me as well.

That's why I think should try out this approach and measure the performance impact. And then take a decision based on the data.

**Comment by tim-dlang** [comment]
> That's why I think should try out this approach and measure the performance impact. And then take a decision based on the data.

I did a first test of doing the semantic analysis at the call site, but it is not complete, and this PR is unchanged. I tested it by compiling std.regex with unittests enabled.

Doing the semantic analysis only at the call site increased the runtime by roughly 0.6%. Unfortunately the semantic analysis also needs to be done at the declaration site, so error messages can be shown for functions never used. Doing the semantic analysis both at the declaration site and the call site solves this, but then the runtime of DMD increases by roughly 4.9%.

Another problem is memory consumption: The scope of the function needs to remain valid, so it can be used by the semantic analysis at the call site. In my test DMD consumed 2.1% more memory.

It was only a first test and there could be possible optimizations, or I made mistakes, but I think doing semantic analysis at the call site would not be worth it with those increases in runtime and memory consumption.

File: compiler/src/dmd/expressionsem.d [hunk 8]
```diff
@@ -15056,10 +15062,121 @@ Expression resolveLoc(Expression exp, const ref Loc loc, Scope* sc)
...
+        foreach (ref element; *exp.arguments)
+        {
+            if (element)
+                element = element.resolveLoc(loc, sc);
+        }
+
+        return exp;
+    }
+
+    Expression visitArray(ArrayExp exp)
```
Comment (dkorpel) [inline compiler/src/dmd/expressionsem.d:15122]: This function is not covered by tests

File: compiler/src/dmd/dtemplate.d [hunk 9]
```diff
@@ -5625,7 +5625,11 @@ extern (C++) final class TemplateValueParameter : TemplateParameter
         if (e)
         {
             e = e.syntaxCopy();
-            if ((e = e.expressionSemantic(sc)) is null)
+            const saveInDefaultArg = sc.inDefaultArg;
+            sc.inDefaultArg = true;
```
Comment (dkorpel) [inline compiler/src/dmd/dtemplate.d:5629]: This 'save and restore' pattern is risky, since `sc` may be referenced by siblings. It's safer to push and pop a scope, see: https://github.com/dlang/dmd/pull/15434#discussion_r1270170347

Although in these cases, I don't think it's possible for there to be expression siblings that aren't default args.

File: compiler/src/dmd/parse.d [hunk 10]
```diff
@@ -6949,33 +6949,6 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
         }
     }
 
-    /*****************************************
-     * Parses default argument initializer expression that is an assign expression,
-     * with special handling for __FILE__, __FILE_DIR__, __LINE__, __MODULE__, __FUNCTION__, and __PRETTY_FUNCTION__.
-     */
-    private AST.Expression parseDefaultInitExp()
```
Comment (dkorpel) [inline compiler/src/dmd/parse.d:6956]: I love this removal

File: compiler/src/dmd/parse.d [hunk 11]
```diff
@@ -8152,33 +8125,23 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
...
-                e = new AST.StringExp(loc, s.toDString());
-                nextToken();
-                break;
-            }
+            e = new AST.FileInitExp(loc, EXP.fileFullPath);
+            nextToken();
+            break;
 
         case TOK.line:
-            e = new AST.IntegerExp(loc, loc.linnum, AST.Type.tint32);
```
Comment (dkorpel) [inline compiler/src/dmd/parse.d:8171]: And I love how the tokens aren't resolved in the parser anymore

**Review by dkorpel — APPROVED** [review]
As others have mentioned, this isn't particularly ideal. I also wish we could re-use existing visitor code for this. However, this has immediate value, and I agree with Steven:

> I would hate to delay it because the implementation details are not perfect.

So I think it's good to get this in.

**Comment by tim-dlang** [comment]
I have added a test for ArrayExp and modified the code changing inDefaultArg, so it uses a new scope.

Unfortunately this change first triggered an assert in `Scope.push` for test compilable/test9701.d: The assert checks that the new scope is not the current scope. `Scope.push` was called inside `Scope.startCTFE`, which was called inside `getMessage(DeprecatedDeclaration dd)`. It was trying to use a scope, which was already freed.

I added the call `em.depdecl._scope.setNoFree();` for deprecated enum members, so this does not happen.

**Comment by RazvanN7** [comment]
This PR introduced a regression: https://issues.dlang.org/show_bug.cgi?id=24560

Fixed in: https://github.com/dlang/dmd/pull/16519

File: compiler/src/dmd/dtemplate.d (hunk 9 above)
Comment (dkorpel) [reply compiler/src/dmd/dtemplate.d]: > It's safer to push and pop a scope

Welp, that aged well :laughing:

https://github.com/dlang/dmd/pull/16781
